traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go

```diff
@@ -0,0 +1,389 @@
+		log.Errorln(err)
+		return tc.DBError, tc.SystemError
+	} else if rowsAffected > 1 {
+		err = errors.New("too many ids returned from fed insert")
+		log.Errorln(err)
+		return tc.DBError, tc.SystemError
+	}
+	fed.SetKeys(map[string]interface{}{"id": id})
+
+	return nil, tc.NoError
+}
+
+func (fed *TOCDNFederation) Read(parameters map[string]string) ([]interface{}, []error, tc.ApiErrorType) {
+
+	// Cannot perform query on tenantID while "rows" aren't closed (limitation of
+	// psql), so we need to get the valid tenentIDs ahead of time.
+	tenantIDs, err := tenant.GetUserTenantIDListTx(fed.ReqInfo.Tx.Tx, fed.ReqInfo.User.TenantID)
+	if err != nil {
+		log.Errorf("getting tenant list for user: %v\n", err)
+		return nil, []error{tc.DBError}, tc.SystemError
+	}
+
+	var query string
+	_, id := parameters["id"]
+	queryParamsToQueryCols := map[string]dbhelpers.WhereColumnInfo{
+		// db tag                                         symbol from query
+		"id":          dbhelpers.WhereColumnInfo{Column: "federation.id", Checker: api.IsInt},
+		"cname":       dbhelpers.WhereColumnInfo{Column: "cname", Checker: nil},
+		"ttl":         dbhelpers.WhereColumnInfo{Column: "ttl", Checker: api.IsInt},
+		"description": dbhelpers.WhereColumnInfo{Column: "description", Checker: nil},
+		"xmlId":       dbhelpers.WhereColumnInfo{Column: "xml_id", Checker: nil},
+		"ds_id":       dbhelpers.WhereColumnInfo{Column: "deliveryservice.id", Checker: api.IsInt},
+	}
+	if id {
+		query = selectByID()
+	} else { // searching by name
+		queryParamsToQueryCols["name"] = dbhelpers.WhereColumnInfo{Column: "cdn.name", Checker: nil}
+		query = selectByCDNName()
+	}
+
+	where, orderBy, queryValues, errs := dbhelpers.BuildWhereAndOrderBy(parameters, queryParamsToQueryCols)
+	where, queryValues = dbhelpers.AddTenancyCheck(where, queryValues, "(tenant_id", tenantIDs)
+	where += " OR tenant_id IS NULL)"
+	if len(errs) > 0 {
+		return nil, errs, tc.DataConflictError
+	}
+
+	query += where + orderBy
+	log.Debugln("Query is ", query)
+
+	rows, err := fed.ReqInfo.Tx.NamedQuery(query, queryValues)
+	if err != nil {
+		log.Errorf("querying federations: %v", err)
+		return nil, []error{tc.DBError}, tc.SystemError
+	}
+	defer rows.Close()
+
+	federations := []interface{}{}
+	for rows.Next() {
+
+		if err = rows.StructScan(&fed.CDNFederation); err != nil {
+			log.Errorf("parsing federation rows: %v", err)
+			return nil, []error{tc.DBError}, tc.SystemError
+		}
+		if fed.ID == nil {
+			log.Errorf("unexpected nil id")
+			return nil, []error{tc.DBError}, tc.SystemError
+		}
+
+		// if we are getting by id, there may not be an attached deliveryservice
+		// DeliveryServiceIDs will not be nil itself, due to the struct scan
+		if id && fed.DsId == nil {
+			fed.DeliveryServiceIDs = nil
+		}
+
+		federations = append(federations, fed)
+	}
+
+	// if federations yields "response": []
+	if len(federations) == 0 {
+
+		if id {
+			return nil, []error{tc.TenantDSUserNotAuthError}, tc.ForbiddenError
```

Comment:
So if the request is asking for the federation by ID, and there are no federations with that ID, then the API should return a 404. If it asks by ID and the federation with that ID exists but the tenant doesn't have access to it, the API should return a 401.

If the requests asks for _all_ of the cdn federations (e.g. like the "index" operation), but there are none to return (even when some are filtered out via tenancy), the API should return a 200 with an empty `[]` rather than a 404.

Reply:
The second paragraph is currently how the code works. 👍
Now just have to differentiate between the 401 and 404.